types/prop-types/index.d.ts

```diff
@@ -1,33 +1,62 @@
 // Type definitions for prop-types 15.5
 // Project: https://github.com/reactjs/prop-types
 // Definitions by: DovydasNavickas <https://github.com/DovydasNavickas>
+//                 Ferdy Budhidharma <https://github.com/ferdaber>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.2
+// TypeScript Version: 2.8
 
-export type Validator<T> = (object: T, key: string, componentName: string, ...rest: any[]) => Error | null;
+import { ReactElement, ReactNode } from 'react';
+
+type IsOptional<T> = undefined | null extends T ? true : undefined extends T ? true : null extends T ? true : false
+
+interface Validator<T> {
+    (object: T, key: string, componentName: string, ...rest: any[]): Error | null;
+}
 
 export interface Requireable<T> extends Validator<T> {
     isRequired: Validator<T>;
 }
 
-export type ValidationMap<T> = {[K in keyof T]?: Validator<T> };
+export type ValidationMap<T> = {
+    [K in keyof T]-?: IsOptional<T[K]> extends true
+        ? Requireable<ReactNode extends T[K]
```

Comment:
Is there any way to check within a conditional type if the type is exactly `any`? Trying to do this:
```ts
type ValidationMapOfAny = ValidationMap<{
  anyProp: any
  optionalAnyProp?: any
}> // want it to be: { anyProp: Validator<any>; optionalAnyProp: Requireable<any> }
```

I don't think there is any way to do this so right now I'm opting to make a breaking change such that `PropTypes.any.isRequired` and `PropTypes.node.isRequired` are invalid use cases, since `any` and `node` technically both cover `undefined`

Reply:
If it extends both `never` and `unknown`, it's any. Pre-unknown, `{} | null | undefined` may do a decent `unknown` impression. Though, as always, test.

Reply:
From testing it out, this is what I'm seeing (I believe it's some kind of special short circuiting logic in the compiler?):
```ts
type IsAny<T> = T extends never ? true : false
type test1 = IsAny<any> // boolean
type test2 = IsAny<string> // false
type test3 = IsAny<never> // never
```

Seems like any time `any` is the type being checked in a conditional type, the resolved type becomes the union of both true and false cases.

I don't believe that getting these cases to work is that important though, since I don't think there are any real-world use cases for `PropTypes.any.isRequired` or `PropTypes.node.isRequired` in JS, and TS users won't likely be using `PropTypes.any`

Reply:
If you still care, maybe
```ts
type IsAnyInnrer<T> = T extends never ? true : false
type IsAny<T> = [boolean] extends [IsAnyInnrer<T>] ? true : false;
type test1 = IsAny<any> // true, I hope
```
could work? (Not in a place where I can check right now) It's probably still best avoided - a case specific to one exact type instead of a class thereof (even any) is a bit off-putting.

Reply:
Good news: it works, it doesn't even need to be boxed as a tuple type, this works:
```ts
type IsAny<T> = boolean extends (T extends never ? true : false) ? true : false
```

I haven't decided yet whether it's worth it to try to capture the `any` type specifically though, for ease of maintenance in the future.